### ISSUE TITLE
[Improvement]: Integrate table semtype into jBallerina type engine

### ISSUE THREAD

**Issue by lochana-chathura** [ballerina-platform/ballerina-lang#42967]
### Description

$subject.

**Comment by lochana-chathura** [comment]
Having a single test failure for the following narrowed-down sample. Looking into it now.

```bal
type H int|string|H[]|map<H>|table<map<H>>|record { H a;};

public function foo() {
     record { H a; } res = { a : "r"};
     anydata x = res.a; // Gives incompatible types: expected 'anydata', found 'H'
}
```

**Comment by lochana-chathura** [comment]
```bal
type H int|string|H[]|map<H>|table<map<H>>|record { H a; };

// H<:A
type A anydata;
```
the above is passing in semtype module unit tests. Hence, the problem is with integration logic.

**Comment by lochana-chathura** [comment]
Problem was same as, https://github.com/ballerina-platform/ballerina-lang/issues/42811#issuecomment-2154461583

**Comment by lochana-chathura** [comment]
Having a test failure at langlib level after rebasing on top of the object integration PR. Looking into it now.

**Comment by lochana-chathura** [comment]
> Having a test failure at langlib level after rebasing on top of the object integration PR. Looking into it now.

The above failure for the regex langlib was causing due to https://github.com/ballerina-platform/ballerina-lang/issues/43161#issuecomment-2246965330

**Comment by lochana-chathura** [comment]
However, after bypassing distinct resolution for objects I am still facing with a subtyping failure for query lang lib.
I'm getting the following error 
```
incompatible types: expected 'ballerina/lang.value:0.0.0:Cloneable', found 'anydata'
```

**Comment by lochana-chathura** [comment]
It seems the semtype resolved for BAnydata from the union members is wrong. It is not equivalent to `Core.createAnydata(cx)`.

The resolved semtype for Cloneable is correct. It is equivalent to the following.
```java
    public static SemType createCloeanble(Context context) {
        SemType memo = context.cloneableMemo;
        Env env = context.env;

        if (memo != null) {
            return memo;
        }
        ListDefinition listDef = new ListDefinition();
        MappingDefinition mapDef = new MappingDefinition();
        SemType tableTy = TableSubtype.tableContaining(env, mapDef.getSemType(env));
        SemType ad = union(VAL_READONLY, union(XML, union(listDef.getSemType(env), union(tableTy,
                mapDef.getSemType(env)))));
        listDef.defineListTypeWrapped(env, ad);
        mapDef.defineMappingTypeWrapped(env, new ArrayList<>(), ad);
        context.cloneableMemo = ad;
        return ad;
    }
```

**Comment by lochana-chathura** [comment]
Sorted out the root cause. The issue was with table resolving logic where I handled the negative table type  definition `type T table<T>`. Previously no tests failed because in the `IMPLEMENTED_TYPES` we had not included the `CELL`.

**Comment by lochana-chathura** [comment]
Having a single test failure. Looking into it. 

Narrowed sample,
```bal
type tempErrorDetail1 record {
    string? name;
};

public type Any A|B;

type A error<tempErrorDetail1> & distinct error;
type B error<tempErrorDetail1> & distinct error;

function testErrorUnionDetailType() {
    Any? err1 = error A("ERROR", name="testFunction"); // Gives incompatible types: expected 'Any?', found 'A'
}
```

**Comment by lochana-chathura** [comment]
The above is not due to a bug in semtype integration. This is because BIntersectionType's effective type does not include the distinct flag.

**Comment by lochana-chathura** [comment]
Update: Got the Windows and Ubuntu build passed. However, Ballerina library 5, 7 & 8 levels are failing.

**Comment by lochana-chathura** [comment]
Fixed the level 7 with [353c8f1](https://github.com/ballerina-platform/ballerina-lang/pull/43092/commits/353c8f13e6dc80561d22cfe68164a7bc01f73032). The failure was due to not handling  `isAssignableIgnoreObjectTypeIds()` where object type ids are ignored explicitly.

Level 5 is still failing. The failing check is the following.
source: 
```
isolated object { public function createInterceptors () returns ([ballerina/http2_tests:2.12.0:DefaultRequestInterceptor,ballerina/http2_tests:2.12.0:DataBindingRequestInterceptor,ballerina/http2_tests:2.12.0:LastRequestInterceptor]); resource function post databinding(string) returns (string); }
```
target: 
```
ballerina/http:2.12.0:InterceptableService
```
The semtype subtyping relation gets the correct result. However, the old path returns with false.

**Comment by lochana-chathura** [comment]
> The semtype subtyping relation gets the correct result. However, the old path returns with false

In the old parser when we call `isAssignable()` recursively, the following fails in the semtype subtyping.

source:
`ballerina/http2_tests:2.12.0:DefaultRequestInterceptor`
target:
`ballerina/http:2.12.0:Interceptor`

**Comment by lochana-chathura** [comment]
Level 7 and 5 compilation failure fixed with [59fd1c5](https://github.com/ballerina-platform/ballerina-lang/pull/43092/commits/59fd1c5613b59bccb3ad46be0eef63e1a6961c3d). 

On top of that, we still ran into failure to huge time consumption by the mappingInhabited algorithm (Issue #43254).
This was fixed with the [1c2b7e7](https://github.com/ballerina-platform/ballerina-lang/pull/43092/commits/1c2b7e71ab5e1a39954ccbc5edc99fd054710418). Now the build is passing.

**Comment by lochana-chathura** [comment]
Closing with #43092
